File: arm/api/forms/img/img.js
// *** *** ***
//
// aon 2022
//
// *** *** ***

window.sovaActions = window.sovaActions || {};
window.sovaActions.img = {
	init2: doc => {
		if (!doc._first) {
			doc.page.dom.current.addEventListener('dragover', e => e.preventDefault());
			doc.page.dom.current.addEventListener('drop', e => window.sovaActions.img.cmd.addFiles(doc, null, e));
			doc._first = true;
			doc.page.forceUpdate();

			let url=`/api/getData?form=${doc.form}&cmd=getFormDir`;
			fetch(url, {method: 'get', credentials: 'include'})
				.then( response => response.text())
				.then( formDir => doc.formDir = formDir)
				.catch( err => doc.msg.error(err.message, 'getFormDir-error:'));
		}

	},
	cmd: {
		btnFiles: () => {
			let el = document.getElementById('inputBgFiles');
			if (el)
				el.click();
		},
		addFiles: (doc, p, e) => {
			e.preventDefault();
			let files = e.dataTransfer?.files || e.target.files;
			let items = e.dataTransfer?.items;
			if (!files && !(items && items[0] && items[0].kind === 'file'))
				return;

			let f = files[0];

	doc.msg.box(`Загрузить файл ${f.name} на сервер?\nsize ${f.size}(${f.type})`, 'Upload')
		.then(() => {
			//if (!f.size)
			//	return doc.msg.ok('Нельзя загрузить файл нулевой длины');

			let formData = new FormData();
			formData.append('bgFile', f);
			
			let par = {
	            method: 'post',
				credentials: 'include',
				body: formData,
			};

	        let csrftoken = doc.util.getCookie('csrftoken');
	        if (csrftoken)
	            par.headers = {'X-CSRFToken': csrftoken};

			let path = doc.formDir ?
				`${doc.formDir}/${doc._path || ''}`
				: 
				doc._path || '';


			let url = `/api/upload?path=${path}`;

			doc.waitLoad = true;
			setTimeout( () => doc.waitLoad && doc.setState({spinner: true}), 500 );

			if (e.target.tagName === 'INPUT')
				e.target.value = '';
		
			fetch(url, par)
				.then( response => response.text())
				.then( text => {
					doc.waitLoad = false;
					doc.setState({spinner: false});
					text !== 'OK' ?
						doc.msg.error(text)
						:
						window.sovaActions.img.cmd.chDir(doc);
				})
				.catch( err => {
					doc.waitLoad = false;
					doc.setState({spinner: false});
					doc.msg.error(err.message, 'Util.doPost-error:');
				});
			})
			.catch(()=>{})
		},

		// ***
		
		oneImg: (doc, img) => {
			let colorPage = doc.mainDoc.sovaPagesByName['settingColors'];
			if (colorPage) {
				let docSP = colorPage.doc;
				let box = docSP.box;
	
				docSP.setField('backgroundImage', img);
				//box.clip.toHist(box, 'tuning', 'backgroundImage');
			}
			else if (doc.fieldValues['VL']) {
				let i = parseInt(doc.fieldValues['VL'], 10);
				let owner = doc.page.owner;
				owner.videoList[i]['image'] = img;
				owner.setField(`vl${i}_image`, img);
				owner.getControl('videoGrid').forceUpdate();
			}
			else {
				let owner = doc.page.owner;
				owner.setField('sticker', img);
				let jsn = [{_teg: 'div', attributes: {style: {
					width:260, 
					height:160, 
					backgroundSize:'100% 100%', 
					backgroundImage:`url('${img}')`,
				}}}];
				owner.setField('stickerImg', jsn);
			}
			doc.page.closePage();
		},
		
		chDir: (doc, path) => {
			doc._path = path;
			
		    let url = `/api/newForm?form=img&list=${doc.getField('list')}&subDir=${path}&vl=${doc.getField('VL')}`;
		    doc.util.jsonByUrl(doc, url)
		        .then( jsn => {
		            doc.setDocProps(jsn);
		            return jsn && jsn.urlForm ? doc.util.jsonByUrl(doc, jsn.urlForm) : null;
		        })
		        .then( page => {
		            doc.register = {};
		            doc.init2 = false;
		            doc.loadForm(page);
		            doc.page.title = `${doc.page.pageName}: ${doc._path}`;
		         })
		        .catch( err => doc.msg.error(err) );
		},
	},
	recalc: {
		BTNLIST: (doc, val) => {
			let url = `/api/newForm?form=img&list=${val}&subDir=${doc._path}&vl=${doc.getField('VL')}`;
		    doc.util.jsonByUrl(doc, url)
		        .then( jsn => {
		            doc.setDocProps(jsn);
		            return jsn && jsn.urlForm ? doc.util.jsonByUrl(doc, jsn.urlForm) : null;
		        })
		        .then( page => {
		            doc.register = {};
		            doc.init2 = false;
		            doc.loadForm(page);
		            doc.page.title = `${doc.page.pageName}: ${doc._path || ''}`;
		         })
		        .catch( err => doc.msg.error(err) );
		},
	}
};

File: arm/api/forms/img/img.py
# -*- coding: utf-8 -*-
'''
AON 2022

'''
from arm.api.forms.formTools import style, _div, _btnD, _img, _input, _field, labell
from arm.api.forms.classPage import Page
from arm.tools.imgHeader import what
from arm.tools.first import err
from arm.api.forms.toolbars import toolbar
from arm.settings import BASE_DIR
from arm.tools.httpMisc import nvResponse

import os

# *** *** ***


class img(Page):
    title = 'Pictures'
    noCaching = True
    dbAlias = 'arm'
    form = 'img'

    def __init__(self, request):
        self.jsCssUrl = '/api/jsv?forms/img/img.js'
        self.pictures = self.pic = os.path.join('static', 'pictures')
        self.formDir = ''
        super().__init__(request)

    # *** *** ***

    def getData(self, dcUK):
        if dcUK.cmd == 'getFormDir':
            return nvResponse(self.formDir)
        else:
            return nvResponse(f'Server error. cmd: {dcUK.cmd}', status=400)
    # *** *** ***

    def page(self, request):
        if request.dcUK.formDir:
            self.formDir = request.dcUK.formDir
            self.pictures = os.path.join(self.pic, self.formDir)
        body = _div(children=[
            *self.picList(request),
            _input(type="file", id='inputBgFiles', onChange='addFiles', multiple=False, **style(display='none'))
        ])
        btnFiles = _btnD('Выбрать и загрузить файл', 'btnFiles', title='Выбрать файл', className='toolbar-button')
        btnList = _field('btnList', 'chb', ['Список'])
        # btnList = _btnD('Список', 'btnList', className='toolbar-button')

        return self.docPage([body], tool=[btnFiles, btnList, toolbar.close_])

    # *** *** ***

    def picList(self, request):  # AnyDesk
        subDir = request.dcUK.subDir
        try:
            path = os.path.join(BASE_DIR, self.pictures, subDir)
            lsfiles = os.listdir(path)
        except Exception:
            try:
                subDir = ''
                path = os.path.join(BASE_DIR, self.pictures)
                lsfiles = os.listdir(path)
            except Exception as ex:
                s = f'404 pages.img: os.listdir-error({path}): {ex}'
                err(s, cat='pages.img')
                return [_div(s)]

        ls = []
        btn = []
        if subDir:
            p = os.path.split(subDir)[0]
            btn.append(
                _btnD('⇑', 'chDir', p, className='chDir', title='вернуться', children=[
                    _div('⇑ ⇑', **style(padding=2)),
                    _div('⇑ ⇑ ⇑', **style(padding=3)),
                    _div('вернуться'),
                ])
            )
            subDir += os.sep

        for k in sorted(lsfiles):
            fil = os.path.join(path, k)
            if os.path.isfile(fil) and what(fil):
                if request.dcUK.list == '1':
                    item = _btnD('', 'oneImg', f'/{self.pictures}/{subDir}{k}', className='chDir',
                        title=f'Выбрать файл "{k}"',
                        **style(margin=5, display='flex'),
                         children=[
                            labell(k, **style(width=200)),
                            _img(src=f'/{self.pictures}/{subDir}{k}', alt=k, **style(width=100, height=100))
                        ],)
                else:
                    item = _btnD('', 'oneImg', f'/{self.pictures}/{subDir}{k}', className='chDir',
                        title=f'Выбрать файл "{k}"',
                        **style(margin=5),
                        children=[
                            _img(src=f'/{self.pictures}/{subDir}{k}', alt=k, **style(width=200, height=200))
                        ],
                    )
                ls.append(item)
            elif os.path.isdir(fil):
                btn.append(
                    _btnD('', 'chDir', f'{subDir}{k}', className='chDir', title='открыть',
                        children=[
                            _img(src=f'/image/folder.png'),
                            _div(k, **style(padding='0 10px'))
                    ])
                )
        if btn:
            ls = [_div(children=btn, **style(textAlign='center', margin=5))] + \
                 [_div(children=ls, **style(textAlign='center', margin=0))]
        return ls

    # *** *** ***

    def queryOpen(self, r):
        dcUK = r.dcUK
        dcUK.doc.vl = dcUK.vl  # index of videoList
        dcUK.doc.subDir = dcUK.subDir
        dcUK.doc.btnlist = 1 if dcUK.list == '1' else ''

    # *** *** ***

